src/cocoa/toga_cocoa/widgets/box.py

```diff
@@ -24,6 +25,10 @@ def create(self):
         self.native = TogaView.alloc().init()
         self.native.wantsLayer = True
 
+        # allow adding subviews
+        self._container = self
+        self.viewport = CocoaViewport(self.native)
```

Comment:
These two lines don't feel right to me. The "container" is a little hazy as a concept, and we'll probably need to formalize it to make widgets like NavigationView work - but even in the usage at present, a Box shouldn't be it's own controller. 

A container is intended as the "top level" widget in any given child heirarchy. You can see this when you set content on a Window; all the children of the window's content widget are set to have the window's content widget as the container. This chain gets interrupted in a couple of distinct cases - usually when there's some "View control" widget in place (e.g., OptionContainer and ScrollContainer).

Widgets also don't automatically get a viewport; the viewport is something that is a high level (i.e., container) level mechanism for getting device DPI independent sizing information on the containing area. 

The best way I can think of to explain containers is to think of the areas that are subject to layout, and areas that fundamentally *define* layout. When you compute layout, you pass in a heirarchy of widgets with styling information - however, something at the top needs to define the amount of space that is actually available. That's the top level window; but some widgets (like SplitContainer) have their own "top level" status, and produce their own sublayouts. 

Does that make sense?

Reply:
It does make sense, I somewhat abused the concepts to get a quick fix.

> You can see this when you set content on a Window; all the children of the window's content widget are set to have the window's content widget as the container.

But they don't. In the example code below, the box does not have a container or a viewport and therefore adding a new widget to it fails. So the real issue seems to be that the viewport and container are not properly propagated to the children.

Reply:
Correct the above: It is only the `container` which is not set properly.

Reply:
I have now reverted the previous misguided commits and added code which properly propagates viewports and containers. According to my understanding:

- If the parent already has a container or viewport, just propagate it.
- If the parent does not have container or viewport, use the window content as container and it's viewport as a viewport. For instance the window content itself *does not* have a container but rather *is* the container for its children).

Because the current implementation only adds children if the parent has a container, it failed to add children to the window content.

A next step would be to support removing children from a box, but that is an API change and will need a separate PR.